### PR TITLE
Upgrade to Bootstrap 4 and Font Awesome.

### PR DESCRIPTION
**Pull request by janinavelasco9**
Updated to Bootstrap 4 and using font awesome and changed the navigation, button/link colors, and in page navigation (nav pills). All pages are still responsive.

### **Pages that were updated:**
- login
- logout
- book
- book success
- calendar
- customers
- services
- users
- settings

### **[Buttons](https://getbootstrap.com/docs/4.0/components/buttons/#examples)**
- Primary buttons are now btn-info 
- Default buttons are now btn-light
This is to keep the UI semi-similar to your existing colors.

### **Links** 
- Now match the primary (btn-info) color.

### **[Inputs](https://getbootstrap.com/docs/4.0/components/forms/#overview)**
Bootstrap updated their input style, now there are borders around the whole field.

### **Header Navigation**
Bootstrap updated their header navigation style, I put my own little spin on it which adds animation when you hover over an item.

### **[Page Navigation](https://getbootstrap.com/docs/4.0/components/navs/#tabs) (previously nav pills)**
Bootstrap updated their nav pills, instead I decided to just use their nav tabs.

Screen Shot 2020-05-05 at 4 07 14 PM
Screen Shot 2020-05-05 at 4 07 29 PM
Screen Shot 2020-05-05 at 4 08 08 PM
Screen Shot 2020-05-05 at 4 08 22 PM
Screen Shot 2020-05-05 at 4 08 28 PM
Screen Shot 2020-05-05 at 4 08 37 PM
Screen Shot 2020-05-05 at 4 08 47 PM
Screen Shot 2020-05-05 at 4 09 12 PM
Screen Shot 2020-05-05 at 4 10 05 PM
Screen Shot 2020-05-05 at 4 10 11 PM

### CODE REVIEW

**Comment by janinavelasco9** [comment]
@alextselegidis - i couldn't get my dev to work in the development branch but fixed any conflicts that i did have. i hope this all still looks good.

**Comment by alextselegidis** [comment]
@janinavelasco9 Awesome thanks, I'll take over from here. 

<small>
<table>
<tr>
<td></td>
<td>&nbsp;&nbsp;Alex Tselegidis, Easy!Appointments Creator <br>
&nbsp;&nbsp;<a href="https://alextselegidis.com">Need a customization? Contact me in person!</a></td>
</tr>
</table>
</small>

**Comment by popod** [comment]
@janinavelasco9 nice work !

Are you able to make the develop branch to work on your side ? Are you able to resolve existing conflicts and to try your work again with a clean code (there is a lot of changes between your base branch and easyappointments/develop) ?

When it will be done, I can take some times to try this PR :)

Thanks

**Comment by popod** [comment]
@janinavelasco9 thanks !

I've tested it and some things are a little buggy.. I think that you should try to make this run, test it and fix some of the big issues..

To make your installation works, you could follow this doc: https://github.com/alextselegidis/easyappointments/tree/develop#setup

For example:

- click on the calendar (backend) cannot open the "new appointment" modal
- `package-lock.json` is corrupted

Some UI are broken:

Capture d’écran 2020-06-09 à 23 26 26
Capture d’écran 2020-06-09 à 23 25 49

And I've this errors in console:

Capture d’écran 2020-06-09 à 23 27 19

But thanks you for your job.. This look really good with Bootstrap 4 !

**Comment by janinavelasco9** [comment]
@popod - i will look into these. i got everything installed and did an npm start, but im not sure what path i should be hitting to load my dev environment.

**Comment by janinavelasco9** [comment]
myself and another engineer tried to get this running locally but we are having issues. we noticed the tables in the DB arent there and dont know what script to run to get them in there so we cant go to the install page cause they are missing. not really sure what to do about this.

**Comment by popod** [comment]
@janinavelasco9 

What is your error (any message, a blank page, ...) ?
In summary, you have to:

1. clone and install

```
Clone your repository
$ git clone https://github.com/janinavelasco9/easyappointments.git

# Go into the repository
$ cd easyappointments

# checkout to your bootstrap-4-and-fontawesome branch
$ git checkout bootstrap-4-and-fontawesome

# delete the corrupted package-lock.json file
$ rm package-lock.json 

# Install dependencies
$ npm install && composer install

# Start the file watcher
$ npm start
```

2. Create the database in mysql
3. Rename the "config-sample.php" file to "config.php" and update its contents based on your environment.

---

When I try with your branch, the install page always display the "loading" item (you have to fix that).. But if you delete it (with the html inspector), you can create the appointment without problems.

It seems to be some error with bootstrap too. You have to remove the `assets/bootstrap/*` folder, download proper bootstrap library from bootstrap website and add it in `assets/ext/bootstrap/*`.

Hope this helps..

**Comment by janinavelasco9** [comment]
@popod - so on the master branch, the mysql database got built with a default inside the container. That is where we are stuck. This develop branch seems so different from the master branch. The master branch was easy to get it up and running. I get how to do the package.json, the composer install, all that...but the database has no tables and i'm not sure how to go about doing that. If it's too much, since i can close this and try and work on it on my free time or when i'm back on the project that i was doing this for.

**Comment by janinavelasco9** [comment]
we did all those steps you said. my backender also tried to get this set up and he asked how to create the mysql database. we went through the docker steps and nothing is working.

**Comment by popod** [comment]
@janinavelasco9 You just have to create the database. All structure and content will be automatically generate with the installation page.

In which step are you blocked exactly ? do you see the install page ? have you a blank page when you trying to display the installation page ?

Your problem can be:
- Incorrect user/password for the appointment database
- If bootstrap is not correctly initialized (follow my latest comment for that), the installation page cannot initialize and the database content cannot be created

What if you try with the develop branch ? I've tried on my side and it works with your branch and the develop branch.

**Comment by janinavelasco9** [comment]
@popod - so let’s forget my branch right now. We have been trying to just get the develop branch working. We are stuck on how to create the database. We cant get to the install page cause there is no table in the database. We need to know HOW to create it. On the master branch when we built it with docker the database built too. We need to know how to do that with develop branch.

**Comment by popod** [comment]
@janinavelasco9 You have to create the mysql database manually. For that, juste follow one of this exemple links (or search on google), but you have to install mysql before on your computer:

- https://dev.mysql.com/doc/refman/8.0/en/creating-database.html
- https://www.mysqltutorial.org/mysql-create-database/

Or you can juste use an online mysql database service like this one:

- https://remotemysql.com

Note that docker seems to be broken with the develop branch..

**Comment by janinavelasco9** [comment]
@popod  - i see. we were trying through docker. i will try with creating it manually. and yes of course. i sent you an invite

**Comment by popod** [comment]
@janinavelasco9 Nice, thanks ! I've add some fix commits to this branch and improve the steps to make this project works in [this comment](https://github.com/alextselegidis/easyappointments/pull/767#issuecomment-644956124).

**Comment by Elte156** [comment]
Hey @popod, I'm the engineer @janinavelasco9 was referring to.
I just ran through the dev branch and got it working. Here's what I had to do.

Add this to the end of `docker/server/Dockerfile`.
Having these dev/build tools in the image will help others be able to have an identical dev env, instead of relying on the separate installs on the host machine.
```Dockerfile
RUN php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" \
    && php -r "if (hash_file('sha384', 'composer-setup.php') === 'e0012edf3e80b6978849f5eff0d4b4e4c79ff1609dd1e613307e16318854d24ae64f26d17af3ef0bf7cfb710ca74755a') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;" \
    && php composer-setup.php --install-dir=/usr/local/bin --filename=composer \
    && php -r "unlink('composer-setup.php');" \
    && apt-get update && apt-get install -y git zip unzip

RUN curl -sL https://deb.nodesource.com/setup_12.x | bash - \
    && apt-get update && apt-get install -y nodejs
```

Copy the sample into `config.php` and change to these details. These match the existing docker-compose file.
```php
    // ------------------------------------------------------------------------
    // GENERAL SETTINGS
    // ------------------------------------------------------------------------

    const BASE_URL      = 'http://localhost:8000';
    const LANGUAGE      = 'english';
    const DEBUG_MODE    = false;

    // ------------------------------------------------------------------------
    // DATABASE SETTINGS
    // ------------------------------------------------------------------------

    const DB_HOST       = 'mysql:3306';
    const DB_NAME       = 'easyappointments';
    const DB_USERNAME   = 'root';
    const DB_PASSWORD   = 'root';
```

Run:
* `docker-compose up`
* `docker exec -it easyappointments-server bash`
* `npm install`
* `composer install`
* Visit <http://localhost:8000> to see install page

When using the docker-compose file, a database is automatically created as `easyappointments`, so there was no need to create it manually.
I believe the issue was the order of operations when initially setting up the config.php and trying to run this with different versions of PHP, NodeJS, and NPM. We should be good now.

**Comment by popod** [comment]
@Elte156 nice ! Happy to heard that this works now :)

I'm not the maintainer of this project.. But I think that you could open a new issue (or reply to this one #819) and propose your fix, it will be great.

**Comment by janinavelasco9** [comment]
@Elte156 - thank you sooo much. i got it up and working. i will start trying to fix some of the bugs in this pr. thank you both!

**Comment by janinavelasco9** [comment]
@popod - i fixed some of the bugs. some of them that you mentioned i didn't see, like the console errors on the calendar page. let me know if there are any other bugs that you see.

File: assets/css/general.css [hunk 1]
```diff
@@ -396,23 +395,6 @@ li.language:hover {
...
-}
-
-a {
-    color: #17a2b8;
-}
-
-a:hover {
-    color: #0a6977;
-}
-
```
Comment (popod) [inline assets/css/general.css]: @janinavelasco9 why did you remove that ? Now in the appointment process, when the "customer informations" form is not valid, the fields are no more marked as red.

**Comment by popod** [comment]
Some others things to improve:

Appointment process for customers
- [ ] in the mini calendar, the round around date numbers are no more center
- [ ] customer informations fields are no more red when invalid (see my last comment)
- [ ] `form-message` id should be in a new line when `#accept-to-terms-and-conditions` checkbox is displayed (you can active this in the settings).

Backend in calendar page
- [ ] when you click on an appointment event, (save, cancel, ...) buttons no more have margins between thems

General
- [ ] it seems that some space between checkbox and it's label are missing
  - settings page, then business logic tab for exemple
  - users page, provider services panel for an other exemple

Nice things
- [ ] it will be nice to change the date and time picker by something better
- [ ] in the appointment process for customers, it will be nice to better display availables periods (hours) by displaying them as `badge` or `button`

**Comment by janinavelasco9** [comment]
Most of this stuff was off of master. So removing those styles was for that branch. I’ll take a look at the bugs. The nice things we can just do in a different PR since this is solely trying to get bootstrap and font awesome working. Thanks!

File: assets/css/general.css (hunk 1 above)
Comment (janinavelasco9) [reply assets/css/general.css]: i must have removed this for something inside the master branch. control-label isnt a thing anymore with bootstrap4. i will add back though and look into it.

**Comment by janinavelasco9** [comment]
@popod  when you click on an appointment event, (save, cancel, ...) buttons no more have margins between them. i see this in my local
Screen Shot 2020-06-19 at 10 55 25 AM
Screen Shot 2020-06-19 at 10 55 16 AM

**Comment by popod** [comment]
@janinavelasco9 I've this in my side (safari and chrome). Have you try to clean your browser cache ?

Capture d’écran 2020-06-19 à 16 58 22

**Comment by janinavelasco9** [comment]
@popod - yupp but i havent tried with different languages so thank you for that. ill take a look :)

**Comment by popod** [comment]
@janinavelasco9 Yes this could be that... with french they never has spaces between buttons. But with english, if I click in an appointment event, they is no spaces too. But in english "Unavailable" events have some space..

**Comment by janinavelasco9** [comment]
@popod - i will look into that more tomorrow. for now, im going to push up some fixes that should solve some of the things you said need to improve

**Comment by janinavelasco9** [comment]
@popod - this should be good i believe. i think i fixed everything you mentioned

**Comment by popod** [comment]
@janinavelasco9 

Yeah you fixed all the items :)

I've found some more problems:

1. In the appointment process, buttons (next and return) are no more full displayed 


2. In the backend calendar, some spaces are missing between `<strong>` title and labels

**Comment by alextselegidis** [comment]
FYI There will be some additional changes to the design, as the app is getting rebranded, the design of some UI elements will need to be updated. 

Thanks for your help so far!

<small>
<table>
<tr>
<td></td>
<td>&nbsp;&nbsp;Alex Tselegidis, Easy!Appointments Creator <br>
&nbsp;&nbsp;<a href="https://alextselegidis.com">Need a customization? Contact me in person!</a></td>
</tr>
</table>
</small>
